libs/ledger/tests/consensus/pos_tests.cpp

```diff
@@ -0,0 +1,258 @@
+  uint64_t const                              block_interval_ms_{5000};
+  NotarisationPtr                             notarisation_;
+};
+
+TEST_F(ConsensusTests, test_valid_block)
+{
+  BlockPtr block = ValidNthBlock(1);
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::YES);
+}
+
+TEST_F(ConsensusTests, test_loose_block)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->block_number++;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_entropy_block_number_mismatch)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->block_entropy.block_number++;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_hash_empty)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->hash    = "";
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_wrong_miner_identity)
+{
+  BlockPtr block  = ValidNthBlock(1);
+  block->miner_id = qual_[1];
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_not_an_aeon)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->block_entropy.confirmations.clear();
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, DISABLED_test_zero_weight)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->weight  = 0;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, DISABLED_test_wrong_weight)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->weight++;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_timestamp_too_early)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block          = ValidNthBlock(2);
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_timestamp_before_previous)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block          = ValidNthBlock(2);
+
+  block->timestamp = 0;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_not_member_of_qual)
+{
+  BlockPtr block = ValidNthBlock(1);
+
+  auto     signer          = std::make_shared<ECDSASigner>();
+  Identity random_identity = Identity{signer->identity().identifier()};
+
+  block->miner_id = random_identity;
+  block->UpdateDigest();
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_mismatched_digest)
+{
+  BlockPtr block = ValidNthBlock(1);
+
+  block->block_entropy.confirmations[ConstByteArray("a")];
+  block->block_entropy.confirmations[ConstByteArray("b")];
+
+  block->UpdateDigest();
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
```

Comment:
Why is the digest mismatching if it is updated after the changes to confirmations?

Reply:
the digest will now be correct for this block, but the signature should be invalid - so I should rename this test. Consensus does not check UpdateDigest is called since this is done when the block is first received